src/main/java/com/aws/greengrass/deployment/DeploymentQueue.java

```diff
@@ -6,76 +6,162 @@
+
+    /**
+     * Map of internal queue id -> deployment instance.
+     */
+    private final Map<String, Deployment> deploymentMap = new HashMap<>();
+
+    /**
+     * <p>If the offered deployment id is unique, then insert the offered deployment at the tail of the queue.</p>
+     *
+     * <p>When an offered deployment has the same deployment id as an already-enqueued element:
+     *     - if the offered deployment meets replacement criteria, then the enqueued element is replaced by the offered
+     *       element, preserving queue order.
+     *     - otherwise, the offered deployment is ignored.</p>
      *
-     * @param deployment deployment
+     * <p>Replacement criteria are as follows:
+     *     - offered.getDeploymentStage().getPriority() > enqueued.getDeploymentStage().getPriority(), OR
+     *     - offered.isCancelled() == true, OR
+     *     - offered.getDeploymentType().equals(SHADOW)</p>
+     *
+     * <p>When an offered deployment is a shadow deployment:
+     *     - if there is already a shadow deployment in the queue, then replace it and preserve queue order.
+     *     - otherwise, enqueue normally.</p>
+     *
+     * @param offeredDeployment the offered deployment instance.
+     * @return true if the queue was modified, otherwise false.
      */
-    public synchronized boolean offer(Deployment deployment) {
-        //For shadow deployment when desired state is reverted, it can result in scheduling a deployment which is
-        // same as
-        if (!DeploymentType.SHADOW.equals(deployment.getDeploymentType()) && deploymentsQueue.contains(deployment)) {
+    public synchronized boolean offer(Deployment offeredDeployment) {
+        if (offeredDeployment == null) {
+            return false;
+        }
+        final String offeredDeploymentInternalId; // the id to employ in the internal queue
+        if (Deployment.DeploymentType.SHADOW.equals(offeredDeployment.getDeploymentType())) {
+            offeredDeploymentInternalId = SHADOW_DEPLOYMENT_QUEUE_ID;
+        } else {
+            offeredDeploymentInternalId = offeredDeployment.getId();
+        }
+
+        // is the internal queue id already in use?
+        final String enqueuedDeploymentId = deploymentIdQueue.stream()
+                .filter(enqueuedId -> enqueuedId.equals(offeredDeploymentInternalId))
+                .findAny()
+                .orElse(null);
+        if (enqueuedDeploymentId != null) {
+            // internal queue id is already in use
+            final Deployment enqueuedDeployment = deploymentMap.get(enqueuedDeploymentId);
+            if (enqueuedDeployment == null) {
+                logger.atError().kv(DEPLOYMENT_ID_LOG_KEY, offeredDeployment.getId())
+                        .kv("InternalQueueId", offeredDeploymentInternalId)
+                        .log("Logic error: internal queue contains id with no corresponding deployment in map");
+                return false;
+            }
```

Comment:
I think you could avoid this condition if you used a LinkedHashMap

Reply:
Yeah, although this nullcheck is purely defensive.  Should never get null unless somebody introduces a bug later.

I like this better than iterating over all LinkedHashMap elements every time I want the last item...

Reply:
I am evidently mistaken about the LinkedHashMap iteration ordering - LinkedHashMap would be better here.